### PR TITLE
#8 issue 61 emulating malloc failures

### PR DESCRIPTION
**Pull request by ioannnikulin**
- To build: `make fulltest`
- Basic syntax of minishell changed, from now on to start an interactive shell, we have to `./minishell --interactive`, to run one command we have to `./minishell --command echo hello world` etc. 
- Target change: now the executable has a numeric parameter `trap` that fails exactly one malloc. This way we can verify that the program exits gracefully wherever a failure happens.
- Side effect: when building with `make fulltest` (which includes `make vania_trapped`), an additional line is printed in the end of the program, which states the total amount of calls of `ft_calloc_if`

To verify: 
- `./minishell --trap 0 --command MOCK_0` no effect, output is still as if there were no trap argument
```
hello world
Total ft_calloc_calls: 30
```
- `./minishell --trap 23 --command MOCK_0` 23 and up to 30 - segmentation fault. This means that we don't have enough protection in the program. Fixing those segfaults will be a separate task.
- `./minishell --trap 31 --command MOCK_0` no effect, ft_calloc_if is called only 30 times

automatic iteration through those values will also be introduced separately

### CODE REVIEW

File: sources/opts.c [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+/*                                                +#+#+#+#+#+   +#+           */
+/*   Created: 2024/11/03 19:06:16 by inikulin          #+#    #+#             */
+/*   Updated: 2024/11/03 19:46:47 by inikulin         ###   ########.fr       */
+/*                                                                            */
+/* ************************************************************************** */
+
+#include "opts.h"
+
+// returns non-0 on error and sets found if matched
+int	opt_debug(const char **act, int *i, t_param *param, int *found)
```
Comment (ioannnikulin) [inline sources/opts.c:16]: @cr3mebrulee

File: sources/param_init.c [hunk 2]
```diff
@@ -47,6 +47,6 @@ t_param	*param_init(void)
 		return (0);
 	if (param_get_cur_dir(param) || param_get_envvars(param))
 		return (0);
-	param->debug_output_level = 0;
+	param->opts.debug_output_level = 0;
```
Comment (ioannnikulin) [inline sources/param_init.c:50]: @cr3mebrulee

Comment (ioannnikulin) [inline sources/param_init.c:50]: @cr3mebrulee

File: sources/opts.c (hunk 1 above)
Comment (ioannnikulin) [inline sources/opts.c:16]: @cr3mebrulee

**Comment by cr3mebrulee** [comment]
Screenshot from 2024-11-15 10-02-44

File: sources/main.c [hunk 3]
```diff
@@ -43,14 +42,19 @@ int	main(int argc, const char **argv)
...
 		return (1);
-	if (argc == 1)
+	if (opts_fill(argc - 1, &argv[1], param))
+		return (finalize(param, 0, 0, 2));
+	pre(param);
+	if (param->opts.interactive)
 		interactive(param);
-	else
-		noninteractive(param, argc - 1, &argv[1]);
+	else if (param->cur_command)
```
Comment (cr3mebrulee) [inline sources/main.c:53]: i dont see, where is` param->cur_command` assigned?
only one place before it happen is `interactive()`, but if `interactive` executed `else if(param->cur_command)` will not.

File: sources/opts_fill.c
Comment (cr3mebrulee) [inline sources/opts_fill.c:1]: do these single command mode replace parsing for developing level?

Comment (ioannnikulin) [reply sources/opts_fill.c:1]: My idea here was to avoid using readline for debug, since it causes a lot of memory leaks

**Review by cr3mebrulee — APPROVED** [review]

**Pull request merged**